core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java

```diff
@@ -195,8 +195,10 @@ private synchronized void init(MasterInquireClient masterInquireClient) {
       mExecutorService
           .submit(new HeartbeatThread(HeartbeatContext.MASTER_METRICS_SYNC, mClientMasterSync,
               (int) Configuration.getMs(PropertyKey.USER_METRICS_HEARTBEAT_INTERVAL_MS)));
-      // register the shutdown hook
-      Runtime.getRuntime().addShutdownHook(new MetricsMasterSyncShutDownHook());
+      // register the shutdown hook if we are not running tests
+      if (!Configuration.getBoolean(PropertyKey.TEST_MODE)) {
+        Runtime.getRuntime().addShutdownHook(new MetricsMasterSyncShutDownHook());
```

Comment:
The issue applies to non-test mode as well - if a user shuts down Alluxio before they exit their client JVM, the JVM will not be able to exit because it's stuck in the heartbeat shutdown hook. If we aren't going to fix the root issue right away, can you file a ticket to make sure it gets addressed before cutting the 1.8 release?

Reply:
As a solution we could not register a JVM hook at all and instead send the final heartbeat when the client is closed.

Reply:
Good point. I think the JVM shutdown hook is still the most reasonable way to ensure this gets called (instead of relying on client apps to close the FileSystem). What do you think about making the metrics heartbeat call a one-shot instead of a retried RPC? It is designed to be used in a heartbeat thread which already schedules it periodically.

Reply:
good point, only JVM hook is likely to actually get called.

That sounds good, or maybe to handle transient network errors we could use a CountingRetry with 2-3 retries. We could still have hangs if the master is very slow, but at least we won't hang just because the master is stopped.